toolchains/solidity/extension/src/gas-estimation.ts

```diff
@@ -0,0 +1,344 @@
+        new vscode.Position(func.line - 1, content.split("\n")[func.line - 1].length)
+      );
+
+      let decoration = { range, renderOptions: { after: { contentText: `    ${gas.toString()} gas` } }  };
+      decorationsArray.push(decoration);
+    }
+  }
+  return decorationsArray;
+}
+
+async function showReport(editor: vscode.TextEditor, reports: ReportDecorators, reportsSaved: ReportDecorators, decorationType: vscode.TextEditorDecorationType) {
+  let report = reports.get(editor.document.uri.path);
+    const reportSaved = reportsSaved.get(editor.document.uri.path);
+
+    if (!report && !reportSaved) {
+    } else if (report && !reportSaved) {
+      editor.setDecorations(decorationType, report);
+    } else if (!report && reportSaved) {
+      editor.setDecorations(decorationType, reportSaved);
+    } else if (report && reportSaved) {
+      report = report.filter((reportElement) => {
+        return !reportSaved.some((reportSavedElement) => {
+          return reportElement.range.isEqual(reportSavedElement.range);
+        });
+      });
+      const combinedReport = [...(report || []), ...(reportSaved || [])];
+      editor.setDecorations(decorationType, combinedReport);
+    }
+}
+
+export function registerGasEstimation() {
+  const decorationType = vscode.window.createTextEditorDecorationType({
+    after: {
+      color: 'rgba(255, 255, 255, 0.5)'
+    }
+  });
+
+  const reports: ReportDecorators  = new Map();
+  let reportsSaved: ReportDecorators  = new Map();
+
+  vscode.workspace.onDidOpenTextDocument(async (document) => {
+    // gas estimate only the main contracts
+    if (document.uri.path.includes("lib") || document.uri.path.includes("test") || document.uri.path.includes("script") || document.uri.path.includes(".git")) {
+      return;
+    }
+    const report = await gasReport(document.getText(), document.uri.path);
+    reports.set(document.uri.path, report);
+
+    vscode.window.visibleTextEditors.forEach((editor) => {
+      showReport(editor, reports, reportsSaved, decorationType);
+    });
+  });
+
+  vscode.workspace.onDidSaveTextDocument(async (document) => {
+    if (document.uri.path.includes("lib") || document.uri.path.includes("test") || document.uri.path.includes("script") || document.uri.path.includes(".git")) {
+      return;
+    }
+    const report = await gasReport(document.getText(), document.uri.path);
+    reports.set(document.uri.path, report);
+
+    vscode.window.visibleTextEditors.forEach((editor) => {
+      showReport(editor, reports, reportsSaved, decorationType);
+    });
+  });
+
+  vscode.window.onDidChangeVisibleTextEditors(async (editors) => {
+    editors.forEach((editor) => {
+      showReport(editor, reports, reportsSaved, decorationType);
+    });
+  });
+
+  vscode.window.onDidChangeActiveTextEditor(async (editor) => {
+    if (editor) {
+      showReport(editor, reports, reportsSaved, decorationType);
+    }
+  });
+
+  vscode.commands.registerCommand(
```

Comment:
This command should be accessible through the UI like in file contextual menu or at the top right of the editor